### PR TITLE
WT-7237 Creating component interface to improve class heirachy in the test framework.

### PR DESCRIPTION
**Pull request by etienneptl**

### CODE REVIEW

File: test/cppsuite/test_harness/test.h [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        return _thread_manager;
+    }
+
+    static const std::string name;
+    static const std::string default_config;
+
+    private:
+    std::vector<component *> _components;
+    configuration *_configuration;
+    std::mutex mtx;
```
Comment (etienneptl) [inline test/cppsuite/test_harness/test.h:153]: This will be removed

File: dist/s_string.ok [hunk 2]
```diff
@@ -742,6 +744,7 @@ enqueue
 enqueued
 enum
 env
+environement
```
Comment (luke-pearson) [inline dist/s_string.ok:747]: This isn't spelled correctly

File: dist/s_string.ok [hunk 3]
```diff
@@ -1415,6 +1418,7 @@ usercfg
 usr
 utf
 util
+utilise
```
Comment (luke-pearson) [inline dist/s_string.ok:1421]: I think the American spelling of utilize already exists and should be used instead

File: dist/s_string.ok [hunk 4]
```diff
@@ -1427,6 +1431,7 @@ vanishingly
 variable's
 variadic
 vectorized
+verifcation
```
Comment (luke-pearson) [inline dist/s_string.ok:1434]: Again spelled incorrectly

File: dist/s_string.ok [hunk 5]
```diff
@@ -733,6 +734,7 @@ eg
 egrep
 emp
 encodings
+encompases
```
Comment (etienneptl) [inline dist/s_string.ok:737]: Spelling will be fixed

File: test/cppsuite/test_harness/conn_api.h [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+#define CONN_API_H
+
+/* Define common resource access functions. */
+namespace test_harness {
+
+WT_CONNECTION *_conn_conn = nullptr;
+std::mutex _conn_mutex;
+
+static void
+conn_api_close()
```
Comment (luke-pearson) [inline test/cppsuite/test_harness/conn_api.h:39]: Todo: Make singleton connection_manager.h

**Comment by luke-pearson** [comment]
@tammybailey, @quchenhao This PR grew a bit from its original scope but on the whole @etienneptl and I agree that this an overall improvement in class layout.

The key changes are:
- Introduction of component.h which defines a load(), run() and finish(). For phase handling, we may extend this later but for now that is sufficient.
- Introduction of connection.h which is a singleton that owns the database connection and subsequent API calls.
- Removal of thread context from the thread_manager, now it only holds threads and contexts are created and destroyed by their owner.

Let me know if you have any questions.

File: test/cppsuite/test_harness/component.h [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+namespace test_harness {
+/*
+ * A component is a class that defines 3 unique stages in its life-cycle, the stages are expected to
+ * be run in sequence.
+ */
+class component {
+    public:
+    /*
+     * The load function should perform all tasked required to setup the component for the main
```
Comment (quchenhao) [inline test/cppsuite/test_harness/component.h:40]: tasks

File: test/cppsuite/test_harness/connection.h [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+extern "C" {
+#include "test_util.h"
+#include "wiredtiger.h"
+}
+
+#include "api_const.h"
+#include "debug_utils.h"
+
+namespace test_harness {
```
Comment (quchenhao) [inline test/cppsuite/test_harness/connection.h:42]: I'm a bit concerned we define connection as a singleton. I think it is a valid use case to operate on two different databases. e.g., one is a copy of the original database. We do this a lot in the python test already.

File: test/cppsuite/test_harness/connection.h [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+        /* Create the working dir. */
+        testutil_make_work_dir(home.c_str());
+
+        /* Open conn. */
+        testutil_check(wiredtiger_open(home.c_str(), NULL, CONNECTION_CREATE, &_conn));
+    }
+
+    WT_SESSION *
+    get_session()
```
Comment (quchenhao) [inline test/cppsuite/test_harness/connection.h:85]: Should this be open_session instead?

File: test/cppsuite/test_harness/test.h [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+    test(const std::string &config, bool enable_tracking)
+    {
+        _configuration = new configuration(name, config);
+        _workload_generator = new workload_generator(_configuration, enable_tracking);
+        _runtime_monitor = new runtime_monitor();
+        _timestamp_manager = new timestamp_manager();
+        _thread_manager = new thread_manager();
+        /*
+         * Ordering is not important here, any dependencies between components should be resolved
+         * internally by the components.
```
Comment (quchenhao) [inline test/cppsuite/test_harness/test.h:63]: What do you mean by internally? How that is supposed to work?

File: test/cppsuite/test_harness/timestamp_manager.h [hunk 11]
```diff
@@ -26,38 +26,23 @@
...
-#include "configuration_settings.h"
-#include "wiredtiger.h"
-#include "wt_internal.h"
+#ifndef TIMESTAMP_MANAGER_H
+#define TIMESTAMP_MANAGER_H
 
 namespace test_harness {
-class test {
+/*
+ * The timestamp monitor class managers global timestamp state for all components in the test
```
Comment (quchenhao) [inline test/cppsuite/test_harness/timestamp_manager.h:34]: manages

**Review by quchenhao — COMMENTED** [review]
The title of the pr is not very clear.

File: test/cppsuite/test_harness/connection.h (hunk 8 above)
Comment (luke-pearson) [reply test/cppsuite/test_harness/connection.h:42]: I think that's a valid concern, so some possible solutions are:
- An interface connection_manager which is implemented by the test and gives it a single connection instances, then the connection manager can be passed to components that may need a connection.
- A connection wrapper which is just passed between objects

If we do end up having multiple connections then I think we'll have other issues with regards to knowing which URI is for which connection etc. The main reason for the singleton was to avoid concurrency issues with accessing connection API calls.

File: test/cppsuite/test_harness/connection.h (hunk 9 above)
Comment (etienneptl) [reply test/cppsuite/test_harness/connection.h:85]: It is true that the title could be misleading, we thought about it with Luke. However the user does not always need to know what's happening behind the scenes. We should either add an explicit comment in the function header or change the title.

File: test/cppsuite/test_harness/test.h (hunk 10 above)
Comment (luke-pearson) [reply test/cppsuite/test_harness/test.h:63]: A component shouldn't be built that it has a dependency on any other component, however if it does it needs to assert that that dependency is met prior to doing what it was planning on doing. I'm hoping we can avoid controlling ordering.

File: test/cppsuite/test_harness/connection.h (hunk 9 above)
Comment (luke-pearson) [reply test/cppsuite/test_harness/connection.h:85]: Sure 👍

File: test/cppsuite/test_harness/workload_generator.h [hunk 12]
```diff
@@ -149,20 +149,24 @@ class workload_generator : public component {
...
 
-        // Set up work
-        operation = context.get_thread_operation();
+        session = connection::instance().get_session();
 
-        // Main loop
+        /* Main loop for worker threads. */
         while (context.is_running()) {
-            switch (operation) {
+            switch (context.get_thread_operation()) {
```
Comment (etienneptl) [inline test/cppsuite/test_harness/workload_generator.h:159]: It should be optimised, but try to avoid to call functions in a while loop. If not optimised, it will add overhead and cause perf issues.

File: test/cppsuite/test_harness/workload_generator.h [hunk 13]
```diff
@@ -149,20 +149,24 @@ class workload_generator : public component {
...
             case thread_operation::INSERT:
                 /* Sleep until it is implemented. */
                 std::this_thread::sleep_for(std::chrono::seconds(1));
                 break;
             case thread_operation::READ:
-                read_operation(context);
+                /*
+                 * This is a bit weird because read operation itself loops on the context running.
+                 * Will be improved later.
```
Comment (etienneptl) [inline test/cppsuite/test_harness/workload_generator.h:167]: You could remove the while loop from the read_operation function.
But opening too many cursors might cause an issue. They could be saved in the thread context and reused.

File: test/cppsuite/test_harness/workload_generator.h (hunk 12 above)
Comment (luke-pearson) [reply test/cppsuite/test_harness/workload_generator.h:159]: What if we expose the thread operation as a public const variable?

File: test/cppsuite/test_harness/workload_generator.h (hunk 13 above)
Comment (luke-pearson) [reply test/cppsuite/test_harness/workload_generator.h:167]: Yeah we'll need to save the cursors in the thread context which is what I had thought about doing in the future, can do in this change.

File: test/cppsuite/test_harness/connection.h (hunk 8 above)
Comment (quchenhao) [reply test/cppsuite/test_harness/connection.h:42]: I don't understand how singleton solves the concurrency issue. I see we are using a mutex in session to control concurrency. How is the singleton related then?

Comment (luke-pearson) [reply test/cppsuite/test_harness/connection.h:42]: Well actually yeah you're correct, the solution to the concurrency issue would be the same even if we passed a reference to a class that held a connection. It would still have a single mutex. The singleton was a lazy way of avoid dependency injection. What would you prefer?

Comment (quchenhao) [reply test/cppsuite/test_harness/connection.h:42]: I think we shouldn't make it a singleton and we should just pass its reference around to whichever class needs it.

File: test/cppsuite/test_harness/workload_generator.h (hunk 12 above)
Comment (etienneptl) [reply test/cppsuite/test_harness/workload_generator.h:159]: We could, but all other variables are private and have getters, seems inconsistent. I'd usually avoid public variables to keep encapsulation.

File: test/cppsuite/test_harness/connection.h (hunk 8 above)
Comment (luke-pearson) [reply test/cppsuite/test_harness/connection.h:42]: Okay well Etienne and I would perter to keep the singleton for now and if the requirements change we'll either pass Connection container object or similar.

Comment (etienneptl) [reply test/cppsuite/test_harness/connection.h:42]: A singleton could manage different types of connection. Passing the reference around seems to be a lot as it is used in quite a few places already and might be used even more in the next dev.

Comment (luke-pearson) [reply test/cppsuite/test_harness/connection.h:42]: Thinking longer term however if we were to manage multiple connections would we have an array of Connection wrappers? I suppose so, the index into the array could be the usecase for the connection.

File: test/cppsuite/test_harness/connection_manager.h [hunk 14]
```diff
@@ -82,14 +82,14 @@ class connection {
...
 
     WT_SESSION *
-    get_session()
+    create_session()
     {
         WT_SESSION *session;
 
         if (_conn == nullptr) {
-            debug_info("Connection is NULL, did you forget to call conn_api_open ?", _trace_level,
+            debug_info("connection is NULL, did you forget to call conn_api_open ?", _trace_level,
```
Comment (etienneptl) [inline test/cppsuite/test_harness/connection_manager.h:90]: Change conn_api_open to connection_manager::create

File: test/cppsuite/test_harness/connection_manager.h [hunk 15]
```diff
@@ -82,14 +82,14 @@ class connection {
...
               DEBUG_ERROR);
-            testutil_die(EINVAL, "Connection is NULL");
+            testutil_die(EINVAL, "connection_wrap is NULL");
```
Comment (etienneptl) [inline test/cppsuite/test_harness/connection_manager.h:92]: Did you mean "connection" ?

File: test/cppsuite/test_harness/connection.h (hunk 8 above)
Comment (etienneptl) [reply test/cppsuite/test_harness/connection.h]: Yes that's correct, a map/an array, any suitable container to store multiple connections. An ID could be returned upon creation of pre defined IDs as we discussed.

**Review by quchenhao — APPROVED** [review]
lgtm

**Comment by etienneptl** [comment]
Your last fixes look good to me Luke!

File: test/cppsuite/test_harness/connection_manager.h [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+#include "test_util.h"
+#include "wiredtiger.h"
+}
+
+#include "api_const.h"
+#include "debug_utils.h"
+
+namespace test_harness {
+/*
+ * Singleton class owning the database connection, provides access to sessions and any other
```
Comment (tammybailey) [inline test/cppsuite/test_harness/connection_manager.h:44]: Maybe reword this in complete sentences?

**Review by tammybailey — APPROVED** [review]
Just a few comments on comments. LGTM otherwise.

File: test/cppsuite/test_harness/test.h [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+        _components.clear();
+    }
+
+    /*
+     * The primary run function that most tests will be able to utilize without much other code.
+     */
+    void
+    run()
+    {
+        int64_t duration_seconds;
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply test/cppsuite/test_harness/test.h:90]: From now on we'll do this but I won't fix most cases in this PR. 👍
